[LICM] Use DomTreeUpdater version of SplitBlockPredecessors, nfc

caojoshua opened the pull request:
The DominatorTree version is marked for deprecation, so we use the
DomTreeUpdater version. We also update sinkRegion() to iterate over
basic blocks instead of DomTreeNodes. The loop body calls
SplitBlockPredecessors. The DTU version calls
DomTreeUpdater::apply_updates(), which may call DominatorTree::reset().
This invalidates the worklist of DomTreeNodes to iterate over.

The added test would crash if we iterated over DomTreeNodes and called
the SplitBlockPredecessors DTU version.

caojoshua committed:
[BasicBlockUtils] Fix SplitBlockPredecessors incorrect dominator insert

When SplitBlockPredecessors uses DomTreeUpdater, it always assumes that
the newly created BB dominates the old BB. This is not true if the
consumer only passes in a subset of producers. For example:

```
bb0:
  br label %bb2
bb1:
  br label %bb2
bb2: ...
```

If the consumer calls SplitBlockPredecessors on bb2 with only bb0 as the
predecessor:

```
bb0:
  br label %new_bb
new_bb:
  br label %bb2
bb1:
  br label %bb2
bb2: ...
```

In this case, new_bb does NOT dominate bb2 because there is an
alternative path to bb2 from bb1. We fix this by only marking the new BB
as dominating the old BB(bb2) if the old BB has a single predecessor.

SplitBlockPredecessors with DominatorTree is deprecated in favor of
using the function with DomTreeUpdater. We move over a consumer in
LICM::splitPredecessorsOfLoopExit() in this patch, which would have
triggered the error.

caojoshua commented:
Found this issue while working on https://github.com/llvm/llvm-project/pull/97196

caojoshua committed:
Fix formatting

nikic commented on llvm/lib/Transforms/Utils/BasicBlockUtils.cpp:
```diff
@@ -1160,7 +1160,10 @@ static void UpdateAnalysisInformation(BasicBlock *OldBB, BasicBlock *NewBB,
+      if (OldBB->getSinglePredecessor()) {
+        assert(OldBB->getSinglePredecessor() == NewBB);
+        Updates.push_back({DominatorTree::Insert, NewBB, OldBB});
+      }
```
I don't really understand this fix. The updates here are just the CFG changes, and as far as I can tell we do indeed always insert a CFG edge from NewBB to OldBB, so it should be reflected in the updates.

nikic replied:
DTU updates are about edges in the CFG, not the DT. The point of DTU is to determine the correct DT updates from CFG updates, so this kind of reasoning in the caller should never be necessary.

caojoshua replied:
OK, didn't know that about DTU. Actually just tested this and even if I comment out the if statement, it still passes the unit tests. However, if I use the built clang to compile anything, it runs into assertions in LICM. I'll look into it.

caojoshua replied:
OK yeah this isn't right. The issue is actually [in LICM](https://github.com/llvm/llvm-project/blob/3d7af093f3d7b5c025e727277b9cc2b899c2d84e/llvm/lib/Transforms/Scalar/LICM.cpp#L570). If I modify it to iterate over BBs instead of DomTreeNodes, it works as intended. I think the DomTreeNodes are getting modified at some point while modifying the DomTree. The if statement somehow avoids the issue.

I'm having trouble creating a smaller reproducer. But the full issue is in https://buildkite.com/llvm-project/github-pull-requests/builds/88168#01911957-5a9a-4e62-bda6-b2f031368134 (from https://github.com/llvm/llvm-project/pull/97196)

caojoshua committed:
Add reproducing test

caojoshua committed:
Revert "Fix formatting"

This reverts commit a5ec0c30517f110328f2d17d54616facbb448fa0.

caojoshua committed:
Revert "[BasicBlockUtils] Fix SplitBlockPredecessors incorrect dominator insert"

This reverts commit 0caf5b57d24fdf3454fd1a9a13eedffeccba7517.

caojoshua committed:
Auto generate file checks

caojoshua committed:
[LICM] Use DomTreeUpdater version of SplitBlockPredecessors, nfc

The DominatorTree version is marked for deprecation, so we use the
DomTreeUpdater version. We also update sinkRegion() to iterate over
basic blocks instead of DomTreeNodes. The loop body calls
SplitBlockPredecessors. The DTU version calls
DomTreeUpdater::apply_updates(), which may call DominatorTree::reset().
This invalidates the worklist of DomTreeNodes to iterate over.

The added test would crash if we iterated over DomTreeNodes and called
the SplitBlockPredecessors DTU version.

caojoshua committed:
Add a note to the test

caojoshua commented:
I completed reworked this patch. I found the root cause of the bugs after modifying LICM to use the DTU version of SplitBlockPredecessors was that we were iterating over DomTreeNodes while updating the DomTree at the same time. I updated the main PR post.

caojoshua committed:
Fix formatting

arsenm commented on llvm/test/Transforms/LICM/iterate-over-dom-nodes.ll:
```diff
@@ -0,0 +1,889 @@
+}
+
+; uselistorder directives
+uselistorder ptr @_ZNK4llvm16itanium_demangle12OutputBuffer18getCurrentPositionEv, { 1, 0 }
```
I doubt the uselistorder is necessary

caojoshua replied:
Yes, removing this in latest commit

arsenm commented on llvm/test/Transforms/LICM/iterate-over-dom-nodes.ll:
```diff
@@ -0,0 +1,889 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -S -passes='cgscc(devirt<4>(inline,function<eager-inv;no-rerun>(early-cse<memssa>,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;no-sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,loop-mssa(licm<allowspeculation>))))' < %s |  ...
```
Should be able to get this down to one pass?

caojoshua replied:
I don't think so, not for this example. This is the best `reduce_pipeline.py` was able to do. In general this issue has been very hard to reproduce.

nikic replied:
You need to do some manual pipeline cleanup after reduce_pipeline. Things like the cgscc and devirt wrapper are usually not needed, as are most pass options.

nikic replied:
It's usually also possible to use some at least partially inlined IR.

caojoshua replied:
I shortened this line quite a bit. There are still four passes and the devirt. Anything less does not reproduce the error.

arsenm commented on llvm/test/Transforms/LICM/iterate-over-dom-nodes.ll:
```diff
@@ -0,0 +1,889 @@
+; Tests that we should not iterate over DomTreeNodes and modify the DomTree at the same time.
+
+target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-unknown-linux-gnu"
```
Shouldn't need triple or datalayout

caojoshua committed:
Remove uselistorder from test

caojoshua committed:
Reduce passes

nikic commented on llvm/test/Transforms/LICM/iterate-over-dom-nodes.ll:
TBH, I think I'd prefer just not having a test for this than having *this* test. Presumably this is making the stars align in exactly the right way to trigger a crash, but it's probably not going to still be doing that half a year form now when something, somewhere changes in a small way. If there's no way to make the issue more reliably reproducible (e.g. via an extra assert somewhere) then this test isn't adding much, and will likely be a PITA to update in the future for random changes across four passes.

caojoshua replied:
I agree. Its not really testing for the changes in this patch.

caojoshua committed:
Remove test

nikic reviewed (approved):
LGTM

dtcxzyw commented:
I noticed that this patch caused a significant compilation time regression (~7%) when building php.
Reproducer: https://github.com/dtcxzyw/llvm-opt-benchmark/blob/main/bench/php/original/parse_date.ll
with commit 6f3c15163f9e: 
```
perf stat -e instructions:u bin/opt -O3 ../../llvm-opt-benchmark/bench/php/original/parse_date.ll --disable-output

Performance counter stats for 'bin/opt -O3 ../../llvm-opt-benchmark/bench/php/original/parse_date.ll --disable-output':

    16,235,016,831      instructions:u                                                        

       2.118263921 seconds time elapsed

       2.101738000 seconds user
       0.015998000 seconds sys
```
with this patch (commit 0bc98349c841):
```
perf stat -e instructions:u bin/opt -O3 ../../llvm-opt-benchmark/bench/php/original/parse_date.ll --disable-output

 Performance counter stats for 'bin/opt -O3 ../../llvm-opt-benchmark/bench/php/original/parse_date.ll --disable-output':

    44,855,633,670      instructions:u                                                        

       5.125998136 seconds time elapsed

       5.106567000 seconds user
       0.019002000 seconds sys
```

See https://github.com/dtcxzyw/llvm-opt-benchmark/issues/1399 for more details.
Can someone have a look?

nikic commented:
I can't say I'm particularly surprised. Manual DT updates are going to be faster than generic DTU updates, especially in pathological cases. We've observed similar things e.g. with DT updates during unrolling, and ended up replacing DTU with manual updates to address it.

TBH this makes me think that replacing existing uses of SplitBlockPredecessors directly updating DT with DTU is not something we should pursue.

caojoshua commented:
I suppose the [SplitBlockPredecessors DTU implementation](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Utils/BasicBlockUtils.cpp#L1163-L1172) is not that smart. If we have a lot of `Preds` we might have to re-evaluate the DomTree many times. It would be more efficient to just find the common dominator D and insert an edge from D to NewBB.

I think its better to not have separate APIs for DominatorTree and DomTreeUpdater. Maybe we can just call `DTU->getDomTree()->splitBlock()`?
